Add Poetry `1.2.0` announcement (#68)
python-poetry/website: content/blog/2022-08-31-announcing-poetry-1-2-0.md

<!--more-->

Poetry 1.2 boasts a massive list of changes, new features and fixes developed over the course of 2 years, with
contributions from dozens of committers.

If you have a previous version of Poetry installed via the [new installer][installation docs], `pipx` or manually,
getting Poetry 1.2.0 is as easy as:

```bash
$ poetry self update
```

  [installation docs]: {{< relref "docs/1.2#installation" >}}

{{% warning %}}
If you installed Poetry using the deprecated `get-poetry.py`, you will need to migrate to the new installer.
First, uninstall with `get-poetry.py`:

```bash
$ curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python3 - --uninstall
```

Then follow the instructions below to install Poetry 1.2.
{{% /warning %}}

Since there are many changes in Poetry 1.2.0, this post details the changes over the following sections:

- [Breaking changes and major features](#breaking-changes-and-major-features)
- [Other noteworthy changes and features](#other-noteworthy-changes-and-features)
- [Minor changes](#minor-changes)
- [New commands](#new-commands)
- [Changes to existing commands](#changes-to-existing-commands)

> - [New configuration options](#new-configuration-options")

Author: ```suggestion
- [New configuration options](#new-configuration-options)
```

The dependencies declared in `tool.poetry.dependencies` are part of an implicit `main` group.

```toml
[tool.poetry.dependencies]  # The implicit `main` dependency group
httpx = "*"
pendulum = "*"

[tool.poetry.group.test.dependencies]
pytest = "^7.1.0"
pytest-mock = "*"
```

Dependency groups, other than the implicit `main` group, should generally contain additional dependencies that are part
of your development process, as installing them is only possible using Poetry and `poetry install`.

If your project has additional dependencies meant to add additional functionality at runtime, they should be declared
using the ecosystem-standard [extras][extras docs] instead. Extras are supported by
package build and install tools such as `pip`.

[extras docs]: {{< relref "docs/1.2/pyproject#extras" >}}

{{% note %}}
**A note about the `dev-dependencies` section**

Any dependency declared in the legacy `dev-dependencies` section will automatically be added to a `dev` group.
Thus, the following examples are equivalent:

```toml
[tool.poetry.dev-dependencies]
pytest = "^7.1.0"
pytest-mock = "*"
```

```toml
[tool.poetry.group.dev.dependencies]
pytest = "^7.1.0"
pytest-mock = "*"
```

> As the `dev-dependencies` is now deprecated, projects should migrate to the new `group` syntax as soon as possible.

Author: Maybe we could mention that 1.1 will not be able to understand this new syntax?

Reviewer: It's a little redundant, but can't hurt.

$ poetry install --with docs
```

If you only want to install the [**default**, non-grouped dependencies](#dependency-groups) (aka the `main` group), you can do so with the
`--only` option:

```bash
$ poetry install --only main
```

Finally, if you wish to install **only specific groups** of dependencies without installing the `main` group, the
`--only` option can be used to do so:

```bash
$ poetry install --only docs,test
```

#### Removing dependencies from a group

The [`remove`][remove docs] command supports a `--group (-G)` flag to remove packages from a
specific group:

```bash
$ poetry remove mkdocs --group docs
```

[add docs]: {{< relref "docs/1.2/cli#add" >}}
[remove docs]: {{< relref "docs/1.2/cli#remove" >}}
[install docs]: {{< relref "docs/1.2/cli#install" >}}

### Plugin support

Poetry now supports a plugin system to alter or expand functionality.

Example use cases include functionality not desirable to the majority of Poetry users, features out of scope to the main
Poetry project, or specialized functionality specific to a project.

The plugin system is designed to allow use of Poetry in these situations without requiring a custom fork.

#### Using plugins

> Poetry automatically loads all plugins installed into the its environment.

Author: ```suggestion
Poetry automatically loads all plugins installed into its environment.
```

appended to the URL:

```bash
$ poetry add git+https://github.com/myorg/mypackage_with_subdirs.git#subdirectory=subdir
```

Manual editing of the `pyproject.toml` is supported as well. Full documentation, including examples is available
[here][git dep docs].

[pep 508]: https://peps.python.org/pep-0508/

[git dep docs]: {{< relref "docs/1.2/dependency-specification#git-dependencies" >}}

### Single page repository support

Poetry now supports discovering and installing dependencies from the 'single page' style of repository. Some
widely-consumed package are not hosted in a [PEP 503]-complaint repository, but are instead listed on a single HTML
page.

To add a single page repository as a source add it like any other repository:

```bash
$ poetry source add <SOURCE_NAME> <PAGE_URL>

# e.g.
$ poetry source add jax https://storage.googleapis.com/jax-releases/jax_releases.html
```

Full documentation is available [here][single page repo docs].

[single page repo docs]: {{< relref "docs/1.2/repositories#single-page-link-source" >}}

### Synchronizing the environment with the lock file

> To ensure that the environment exactly matches the lockfile, the [`install`][install docs] command has gained a new

Author: ```suggestion
To ensure that the environment exactly matches the lock file, the [`install`][install docs] command has gained a new
```

Previously, configuring sources required manual edits to `pyproject.toml`.

#### `source add`

The [`source add`][source add docs] command adds a new source configuration to `pyproject.toml`:

```bash
$ poetry source add pypi-test https://test.pypi.org/simple/
```

#### `source show`

The [`source show`][source show docs] command displays information on all configured sources for the project:

```bash
$ poetry source show
```

Optionally, you can limit output to one or more sources by specifying them by name:

```bash
$ poetry source show pypi-test
```

#### `source remove`

The [`source remove`][source remove docs] command removes a configured source from `pyproject.toml`:

```bash
$ poetry source remove pypi-test
```

[source docs]: {{< relref "docs/1.2/cli#source" >}}
[source add docs]: {{< relref "docs/1.2/cli#source-add" >}}
[source show docs]: {{< relref "docs/1.2/cli#source-show" >}}
[source remove docs]: {{< relref "docs/1.2/cli#source-remove" >}}

## Changes to existing commands

Poetry 1.2 brings changes to several commands, primarily related to [dependency groups](#dependency-groups).

### Global options

> - Added `--no-cache` to disable all usage of cached packages and source contents

Author: ```suggestion
- Added `--no-cache` to disable all usages of cached packages and source contents
```

  [source remove docs]: {{< relref "docs/1.2/cli#source-remove" >}}

## Changes to existing commands

Poetry 1.2 brings changes to several commands, primarily related to [dependency groups](#dependency-groups).

### Global options

- Added `--no-cache` to disable all usage of cached packages and source contents

### `about`

- Now displays Poetry and poetry-core versions

### `add`

- Added `--editable` to add a dependency in editable mode
- Added `--group` / `-G` to add a dependency to a specific dependency group
- Now understands [PEP 508]-style dependency specifications

### `env remove`

- Added `--all` to delete all environments
- It is now possible to remove multiple environments in one invocation by specifying multiple names

### `install`

- Added `--all-extras` to install all extras
- Added `--only` to strictly install one or more dependency groups (ignoring the implicit `main` group)
- Added `--only-root` to install the project without any dependencies
- Added `--with` to select one or more dependency groups to install in addition to the implicit `main` group
- Added `--without` to skip installing one or more dependency groups

> - Added `--sync` to only ensure installed dependencies exactly match the lockfile (and configured groups)

Author: ```suggestion
- Added `--sync` to ensure installed dependencies exactly match the lock file (and configured groups)
```

- Deprecated `--no-dev` (use `--without dev` instead)
- Deprecated `--remove-untracked` (use `--sync` instead)

### `lock`

- Added `--check` to verify `poetry.lock` agrees with `pyproject.toml`

### `new`

- Added `--readme` to specify the README file format

### `publish`

- Added `--skip-existing` to ignore errors from files already existing in the repository

### `show`

- Added `--only` to strictly select one or more dependency groups to show (ignoring the implicit `main` group)
- Added `--why` to interrogate why a dependency is required (works standalone, or with `--tree` for a single package)
- Added `--with` to select one or multiple dependency groups to show in addition to the implicit `main` group

> - Added `--without` skip selecting one or more dependency groups

Author: ```suggestion
- Added `--without` to skip selecting one or more dependency groups
```
